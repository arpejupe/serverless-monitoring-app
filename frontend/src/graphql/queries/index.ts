export const listEvents = `query GetAllEntries {
    getDataEntries {
      items {
        id
        name
        url
        averageLatencyMs
        lastSample
        status
        logo
      }
    }
  }`;
