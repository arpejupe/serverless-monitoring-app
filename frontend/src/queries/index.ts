export const ListEvents = `query GetAllEntries {
    getDataEntries {
      items {
        id
      }
    }
  }`;
