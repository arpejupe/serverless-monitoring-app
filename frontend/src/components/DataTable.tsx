import { GraphQLResult } from '@aws-amplify/api/lib/types';
import { API, graphqlOperation } from 'aws-amplify';
import React, { useEffect, useState } from 'react';
import {
  Dimmer,
  Header,
  Icon,
  Image,
  Loader,
  Segment,
  Statistic,
  Table,
} from 'semantic-ui-react';
import { ListEvents } from '../queries';

interface IEntry {
  id: string;
  name: string;
  url: string;
  averageLatencyMs: number;
  lastSample: string;
  status: string;
  logo: string;
}

interface IGraphQlResult {
  getDataEntries: { items: IEntry[] };
}

const SiteStatus = ({ status }: { status: string }) => {
  if (status === 'PASS') {
    return (
      <Table.Cell positive={true}>
        <Icon name="checkmark" />
        Passing
      </Table.Cell>
    );
  } else if (status === 'WARNING') {
    return (
      <Table.Cell warning={true}>
        <Icon name="attention" />
        Warning
      </Table.Cell>
    );
  } else if (status === 'ERROR') {
    return (
      <Table.Cell error={true}>
        <Icon name="close" />
        Error
      </Table.Cell>
    );
  } else {
    return null;
  }
};

const Row = (props: IEntry) => {
  const { averageLatencyMs, id, lastSample, logo, name, status, url } = props;
  const lastSampleDate = new Date(lastSample);
  return (
    <Table.Row key={id}>
      <Table.Cell>
        <Header as="h4" image={true}>
          <Image src={logo} />
          <Header.Content>{name}</Header.Content>
        </Header>
      </Table.Cell>
      <Table.Cell>{url}</Table.Cell>
      <SiteStatus status={status} />
      <Table.Cell>
        <Statistic
          horizontal={true}
          label="ms"
          value={averageLatencyMs.toFixed(2)}
        />
      </Table.Cell>
      <Table.Cell>{lastSampleDate.toString()}</Table.Cell>
    </Table.Row>
  );
};

const DataTable = () => {
  const [loading, setLoading] = useState(true);
  const [entries, setEntries] = useState([] as IEntry[]);

  const fetchData = async () => {
    const { data = { getDataEntries: { items: [] } } } = (await API.graphql(
      graphqlOperation(ListEvents),
    )) as GraphQLResult;

    const result = data as IGraphQlResult;

    setEntries(
      result.getDataEntries.items.sort((a, b) => a.name.localeCompare(b.name)),
    );
    setLoading(false);
  };

  useEffect(() => {
    const interval = setInterval(fetchData, 1000);
    return () => clearInterval(interval);
  }, []);

  return (
    <Segment loading={loading} basic={true}>
      <Table celled={true}>
        <Table.Header>
          <Table.Row>
            <Table.HeaderCell>Party Name</Table.HeaderCell>
            <Table.HeaderCell>Site Address</Table.HeaderCell>
            <Table.HeaderCell>Site Status</Table.HeaderCell>
            <Table.HeaderCell>Average Latency</Table.HeaderCell>
            <Table.HeaderCell>Last Sample</Table.HeaderCell>
          </Table.Row>
        </Table.Header>

        <Table.Body>{entries.map(Row)}</Table.Body>
      </Table>
    </Segment>
  );
};

export default DataTable;
