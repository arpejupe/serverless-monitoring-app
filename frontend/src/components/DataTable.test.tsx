import React from 'react';
import renderer from 'react-test-renderer';
import { render, waitForElement } from 'react-testing-library';
import { ListEvents } from '../queries';
import DataTable, { IEntry } from './DataTable';

jest.mock('aws-amplify');
jest.useFakeTimers();

describe('DataTable', () => {
  beforeEach(() => {
    jest.clearAllMocks();
  });

  test('renders without crashing', () => {
    const { getByText } = render(<DataTable />);
    expect(getByText('Party Name')).toBeInTheDocument();
  });

  test('matches snapshot', () => {
    const tree = renderer.create(<DataTable />).toJSON();
    expect(tree).toMatchSnapshot();
  });

  test('updates data from graphql api', async () => {
    const { API, graphqlOperation } = require('aws-amplify');

    const entries: IEntry[] = [
      {
        averageLatencyMs: -1,
        id: 'id1',
        lastSample: new Date('1947/1/1').toISOString(),
        logo: 'logo1',
        name: 'name1',
        status: 'ERROR',
        url: 'url1',
      },
      {
        averageLatencyMs: 3000,
        id: 'id2',
        lastSample: new Date('1948/1/1').toISOString(),
        logo: 'logo2',
        name: 'name2',
        status: 'WARNING',
        url: 'url2',
      },
      {
        averageLatencyMs: 1000,
        id: 'id3',
        lastSample: new Date('1949/1/1').toISOString(),
        logo: 'logo3',
        name: 'name3',
        status: 'PASS',
        url: 'url3',
      },
      {
        averageLatencyMs: 4000,
        id: 'id4',
        lastSample: new Date('1950/1/1').toISOString(),
        logo: 'logo4',
        name: 'name4',
        status: 'Unknown',
        url: 'url4',
      },
    ];
    const results = { data: { getDataEntries: { items: entries } } };
    API.graphql.mockReturnValue(results);

    const operation = {};
    graphqlOperation.mockReturnValue(operation);

    const { container, getByTestId } = render(<DataTable />);

    expect(setInterval).toHaveBeenCalledTimes(1);
    expect(setInterval).toHaveBeenCalledWith(expect.any(Function), 1000);

    jest.advanceTimersByTime(1500);

    jest.useRealTimers();

    await waitForElement(() => getByTestId(entries[0].id));

    expect(API.graphql).toHaveBeenCalledTimes(1);
    expect(API.graphql).toHaveBeenCalledWith(operation);

    expect(graphqlOperation).toHaveBeenCalledTimes(1);
    expect(graphqlOperation).toHaveBeenCalledWith(ListEvents);

    expect(container.firstChild).toMatchSnapshot();
  });
});
