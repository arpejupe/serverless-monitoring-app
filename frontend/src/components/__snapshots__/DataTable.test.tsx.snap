// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DataTable matches snapshot 1`] = `
<div
  className="ui basic loading segment"
>
  <table
    className="ui celled table"
    data-testid="data-table"
  >
    <thead
      className=""
    >
      <tr
        className=""
      >
        <th
          className=""
        >
          Party Name
        </th>
        <th
          className=""
        >
          Site Address
        </th>
        <th
          className=""
        >
          Site Status
        </th>
        <th
          className=""
        >
          Average Latency
        </th>
        <th
          className=""
        >
          Latest Sample
        </th>
      </tr>
    </thead>
    <tbody
      className=""
    />
  </table>
</div>
`;

exports[`DataTable updates data from graphql api 1`] = `
<div
  class="ui basic segment"
>
  <table
    class="ui celled table"
    data-testid="data-table"
  >
    <thead
      class=""
    >
      <tr
        class=""
      >
        <th
          class=""
        >
          Party Name
        </th>
        <th
          class=""
        >
          Site Address
        </th>
        <th
          class=""
        >
          Site Status
        </th>
        <th
          class=""
        >
          Average Latency
        </th>
        <th
          class=""
        >
          Latest Sample
        </th>
      </tr>
    </thead>
    <tbody
      class=""
    >
      <tr
        class=""
        data-testid="id1"
      >
        <td
          class=""
        >
          <h4
            class="ui image header"
          >
            <img
              class="ui image"
              src="logo1"
            />
            <div
              class="content"
            >
              name1
            </div>
          </h4>
        </td>
        <td
          class=""
        >
          url1
        </td>
        <td
          class="error"
        >
          <i
            aria-hidden="true"
            class="close icon"
          />
          Error
        </td>
        <td
          class=""
        >
          <div
            class="ui horizontal statistic"
          >
            <div
              class="value"
            >
              -1.00
            </div>
            <div
              class="label"
            >
              ms
            </div>
          </div>
        </td>
        <td
          class=""
        >
          Wed Jan 01 1947 00:00:00 GMT+0000 (GMT)
        </td>
      </tr>
      <tr
        class=""
        data-testid="id2"
      >
        <td
          class=""
        >
          <h4
            class="ui image header"
          >
            <img
              class="ui image"
              src="logo2"
            />
            <div
              class="content"
            >
              name2
            </div>
          </h4>
        </td>
        <td
          class=""
        >
          url2
        </td>
        <td
          class="warning"
        >
          <i
            aria-hidden="true"
            class="attention icon"
          />
          Warning
        </td>
        <td
          class=""
        >
          <div
            class="ui horizontal statistic"
          >
            <div
              class="value"
            >
              3000.00
            </div>
            <div
              class="label"
            >
              ms
            </div>
          </div>
        </td>
        <td
          class=""
        >
          Thu Jan 01 1948 00:00:00 GMT+0000 (GMT)
        </td>
      </tr>
      <tr
        class=""
        data-testid="id3"
      >
        <td
          class=""
        >
          <h4
            class="ui image header"
          >
            <img
              class="ui image"
              src="logo3"
            />
            <div
              class="content"
            >
              name3
            </div>
          </h4>
        </td>
        <td
          class=""
        >
          url3
        </td>
        <td
          class="positive"
        >
          <i
            aria-hidden="true"
            class="checkmark icon"
          />
          Passing
        </td>
        <td
          class=""
        >
          <div
            class="ui horizontal statistic"
          >
            <div
              class="value"
            >
              1000.00
            </div>
            <div
              class="label"
            >
              ms
            </div>
          </div>
        </td>
        <td
          class=""
        >
          Sat Jan 01 1949 00:00:00 GMT+0000 (GMT)
        </td>
      </tr>
      <tr
        class=""
        data-testid="id4"
      >
        <td
          class=""
        >
          <h4
            class="ui image header"
          >
            <img
              class="ui image"
              src="logo4"
            />
            <div
              class="content"
            >
              name4
            </div>
          </h4>
        </td>
        <td
          class=""
        >
          url4
        </td>
        <td
          class=""
        >
          Unknown
        </td>
        <td
          class=""
        >
          <div
            class="ui horizontal statistic"
          >
            <div
              class="value"
            >
              4000.00
            </div>
            <div
              class="label"
            >
              ms
            </div>
          </div>
        </td>
        <td
          class=""
        >
          Sun Jan 01 1950 00:00:00 GMT+0000 (GMT)
        </td>
      </tr>
    </tbody>
  </table>
</div>
`;
